### PR TITLE
Provide friendlier access to request variants

### LINKED ISSUE #18933: Better access to request.variant
Currently, `variant=` sets an ivar with an array of the options. So request.variant will return something like `[ :phone ]` if you set `request.variant = :phone`. That's cumbersome to deal with on query. You now have to do either `request.variant.first == :phone` or you have to do `request.variant == [ :phone ]`. It gets even worse if you set multiple variants, like `request.variant = [ :tablet, :phone ]`. Checking if either is in there is a hassle.

So, let's wrap that variant in a better access class, so the following api works:

```ruby
request.variant = :phone
request.variant.phone? # => true
request.variant.whatever? # => false

request.variant = [ :tablet, :phone ]
request.variant.tablet? # => true
request.variant.phone? # => true
request.variant.any?(:tablet, :phone) => true
request.variant.any?(:tablet, :desktop) => true
request.variant.any?(:watch, :desktop) => false
```

### PR DESCRIPTION
**Pull request by georgeclaghorn**
Currently, `request.variant` is a dumb reader for the underlying array used to store the request's variants. If you set the variant like so:

``` ruby
request.variant = :phone
```

...then `request.variant` will return `[:phone]`. That can be cumbersome to deal with, since you have to check `request.variant.first == :phone`, `request.variant == [:phone]`, `request.variant.to_a.include?(:phone)`, etc.

This PR implements a friendlier interface:

``` ruby
request.variant = :phone
request.variant.phone?   # true
request.variant.tablet?  # false

request.variant = [:phone, :tablet]
request.variant.phone?    # true
request.variant.tablet?   # true
request.variant.desktop?  # false
request.variant.any?(:phone, :desktop)  # true
request.variant.any?(:desktop, :watch)  # false
```

Fixes #18933.

### CODE REVIEW

File: actionpack/lib/action_dispatch/http/mime_negotiation.rb [hunk 1]
```diff
@@ -139,6 +141,33 @@ def negotiate_mime(order)
...
+
+        def initialize(variants)
+          @variants = variants
+        end
+
+        def any?(*candidates)
+          (variants & candidates).any?
+        end
+
+        def to_ary
```
Comment (kaspth) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: What is it this gets us?

File: actionpack/lib/action_dispatch/http/mime_negotiation.rb [hunk 2]
```diff
@@ -139,6 +141,33 @@ def negotiate_mime(order)
         order.include?(Mime::ALL) ? format : nil
       end
 
+      class VariantInquirer # :nodoc:
```
Comment (simi) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: What about to introduce `ArrayInquirer` to `ActiveSupport`? There is already [similar code](http://guides.rubyonrails.org/active_support_core_extensions.html#inquiry) for `String`.

File: actionpack/lib/action_dispatch/http/mime_negotiation.rb [hunk 3]
```diff
@@ -75,10 +75,12 @@ def formats
 
       # Sets the \variant for template.
       def variant=(variant)
-        if variant.is_a?(Symbol)
-          @variant = [variant]
-        elsif variant.nil? || variant.is_a?(Array) && variant.any? && variant.all?{ |v| v.is_a?(Symbol) }
+        if variant.nil?
           @variant = variant
```
Comment (lucasmazza) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: Now that we have the `VariantInquirer` class, might make sense to initialize the `@variant` ivar with an empty `VariantInquirer` instance, otherwise the client code would have to check if `request.variant` is `nil` before calling the query methods like `phone?` or `tablet?`.

Comment (kaspth) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: I might be mistaken here, but I think these if statements can be simplified as:

```ruby
variants = Array(variants).compact

if variants.all? { |v| v.is_a?(Symbol) }
  @variant = VariantInquirer.new(variants)
else
  raise ArgumentError, "super long message"
end
```

Though I don't know how much of a hot spot this code path is.

Comment (kaspth) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: Ah, `Array(nil)` returns an empty array, so there's no need for the compact call.

Comment (kaspth) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: Just continuing my own ping-pong. It could also be:

```ruby
@variant = VariantInquirer.new Array(variant).tap do |variants|
  unless variants.all? { |v| v.is_a?(Symbol) }
    raise ArgumentError, "super long message"
  end
end
```

Comment (lucasmazza) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: :+1: for the non-tap version

File: actionpack/lib/action_dispatch/http/mime_negotiation.rb (hunk 2 above)
Comment (georgeclaghorn) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: That'd be a nice generalization, but I think it should be accomplished in a separate PR. As it is, this is specific to the array of Symbols used to store variants.

File: actionpack/lib/action_dispatch/http/mime_negotiation.rb (hunk 1 above)
Comment (georgeclaghorn) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: This allows `request.variant` to be compared to other arrays, since Array comparison methods like #==, #&, and #|  call `to_ary` on the other array:

```
request.variant = :phone
# => #<VariantInquirer @variants=[:phone]>

[:phone] == request.variant
# => true
```

Existing framework code [relies on this behavior](https://github.com/rails/rails/blob/676d4ae787254e7842834e31204e5b7350261bbd/actionpack/lib/action_controller/metal/mime_responds.rb#L293) in a few places.

File: actionpack/lib/action_dispatch/http/mime_negotiation.rb (hunk 3 above)
Comment (georgeclaghorn) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: Done.

File: actionpack/lib/action_dispatch/http/mime_negotiation.rb [hunk 4]
```diff
@@ -139,6 +141,33 @@ def negotiate_mime(order)
         order.include?(Mime::ALL) ? format : nil
       end
 
+      class VariantInquirer # :nodoc:
+        delegate :==, :each, :empty?, to: :variants
+
+        def initialize(variants)
```
Comment (simi) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: `variants = []` to avoid `VariantInquirer.new([])` and use `VariantInquirer.new` instead

Comment (georgeclaghorn) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: :+1:

File: actionpack/lib/action_dispatch/http/mime_negotiation.rb [hunk 5]
```diff
@@ -139,6 +141,33 @@ def negotiate_mime(order)
...
+        def any?(*candidates)
+          (variants & candidates).any?
+        end
+
+        def to_ary
+          variants
+        end
+
+        private
+          attr_reader :variants
```
Comment (simi) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: This should be defined on the top of its scope IMHO.

Comment (georgeclaghorn) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: IMO, it shouldn't. It's intended for internal use only. Doesn't make sense to call `request.variant.variants`.

File: actionpack/lib/action_dispatch/http/mime_negotiation.rb [hunk 6]
```diff
@@ -75,18 +73,22 @@ def formats
...
-          @variant = [variant]
-        elsif variant.nil? || variant.is_a?(Array) && variant.any? && variant.all?{ |v| v.is_a?(Symbol) }
-          @variant = variant
+        variant = Array(variant)
+
+        if variant.all? { |v| v.is_a?(Symbol) }
+          @variant = VariantInquirer.new(variant)
         else
-          raise ArgumentError, "request.variant must be set to a Symbol or an Array of Symbols, not a #{variant.class}. " \
+          raise ArgumentError, "request.variant must be set to a Symbol or an Array of Symbols. " \
```
Comment (kaspth) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: I don't know if we can just get rid of outputting the class.

Comment (georgeclaghorn) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: I'll put it back. It seemed awkward that if you set `request.variant` to an array of non-Symbol objects, you'd see:

> request.variant must be set to a Symbol or an Array of Symbols, not a Array.

Comment (kaspth) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: You know what, leave it :smile:

Comment (georgeclaghorn) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: I already put it back. :smile:

Comment (kaspth) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: I know, I saw. Sorry about that. :wink:

File: actionpack/lib/action_dispatch/http/mime_negotiation.rb [hunk 7]
```diff
@@ -139,6 +141,33 @@ def negotiate_mime(order)
...
+
+        def to_ary
+          variants
+        end
+
+        private
+          attr_reader :variants
+
+          def method_missing(name, *args)
+            if name[-1] == '?'
```
Comment (egilburg) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: `name.end_with?('?')`

Comment (georgeclaghorn) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: `name` is a symbol. No #end_with? method.

File: actionpack/lib/action_dispatch/http/mime_negotiation.rb (hunk 5 above)
Comment (carlosantoniodasilva) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: It should be made protected then, private attr accessors raise ruby warnings.

Comment (georgeclaghorn) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: TIL! Fixed.

File: actionpack/lib/action_dispatch/http/mime_negotiation.rb [hunk 8]
```diff
@@ -139,6 +141,34 @@ def negotiate_mime(order)
...
+        def any?(*candidates)
+          (variants & candidates).any?
+        end
+
+        def to_ary
+          variants
+        end
+
+        protected
+          attr_reader :variants
```
Comment (simi) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: I still wonder why this can't be public since you can access it via `to_ary` also.

Comment (georgeclaghorn) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: It shouldn't be public because `request.variant.variants` is an awkward and opaque invocation for accessing the variants as an array.

I'd be open to exposing it publicly as `to_a`.

Comment (simi) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: actually `request.variant.to_a` is weird to me (typing singular to array)

What about to use another name for this array? What about `request.variant.candidates` or something similar?

Comment (georgeclaghorn) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: It doesn't really seem that weird to me. It's an object that can be converted to a plain-old array. Anyway, what's the case where you would need access to the array? I tried to think of one and drew a blank.

File: actionpack/lib/action_dispatch/http/mime_negotiation.rb [hunk 9]
```diff
@@ -139,6 +141,31 @@ def negotiate_mime(order)
         order.include?(Mime::ALL) ? format : nil
       end
 
+      class VariantInquirer # :nodoc:
+        delegate :==, :each, :empty?, to: :@variants
```
Comment (egilburg) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: Perhaps `==` should be delegated to `@variants` in globbed form:

```ruby
def ==(*variants)
  @variants == variants
end
```

Before:
`request.variant == :tablet # will always return false so probably not something user wants`

After:
`request.variant == :tablet` # will return true if tablet is the exact match and no other variants exist.

Comment (egilburg) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: Alternatively if exact match is not something we need in api, then we can just alias `==` to `any?`. IMO either option would be more consistent than just delegating `==` without globbing.

File: actionpack/lib/action_dispatch/http/mime_negotiation.rb [hunk 10]
```diff
@@ -139,6 +141,31 @@ def negotiate_mime(order)
...
+
+        def initialize(variants = [])
```
Comment (egilburg) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: If supporting `==` operator, perhaps cast this to `Set` for order-insensitive comparison.

File: actionpack/lib/action_dispatch/http/mime_negotiation.rb (hunk 9 above)
Comment (georgeclaghorn) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: With globbing:

```ruby
request.variant = [:phone, :tablet]
request.variant == [:phone, :tablet]
# => false
```

Perhaps the solution here is to do something similar to what we do when assigning `request.variant`:

```ruby
def ==(other)
  @variants == Array(other)
end
```

Comment (kaspth) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: I still don't understand why we need #==. The point of this is to provide a better interface, i.e. #any?.

Are there places in the framework that use == with variants?

Comment (georgeclaghorn) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: I thought more tests failed without it, but it seems that only one does, and that's because the test itself uses `assert_equal` to compare `response.variant` with a symbol. I'm fine with removing it.

Comment (georgeclaghorn) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: On the other hand, supporting #== means an easier upgrade to 5.0 for apps that already use it.

Comment (kaspth) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: I don't know if that's commonly used. But we could do a deprecation cycle to be safe.

Comment (georgeclaghorn) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: Couldn't find any public apps on Github using #==, so I just removed it. I don't think it's worth a deprecation cycle.

Comment (kaspth) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: Nice and well done :+1:

Comment (georgeclaghorn) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: For what it's worth, I kept looking and found two apps using #== in the first 15 pages of Ruby code search results on GitHub:

* [codemy/blogmenow](https://github.com/codemy/blogmenow/blob/17a248fd9b3a28dca146e0957b31f90f51237bd7/app/controllers/application_controller.rb)
* [cassiuschen/bdfzer-sso](https://github.com/cassiuschen/bdfzer-sso/blob/2ca76a5aec9e9f9267955202d0c3058496afccd6/app/controllers/home_controller.rb)

Not sure if this changes whether we should leave #== in. It's clearly not widely used, but it makes me nervous that it could silently change behavior for the few apps that do rely on it.

**Comment by egilburg** [comment]
@dhh would an ArrayInquirer also imply an ArrayWithIndifferentAccess? or do a runtime cast `to_s` or `to_str`?

e.g. in this example:

```ruby
class SoftString
  def to_s
    'soft'
  end
end

class HardString
  def to_str
    'hard'
  end
end

ai = ActiveSupport::ArrayInquirer.new(['foo', :bar', SoftString.new, HardString.new])
```

What should each of these return? true, false, or cast exception?

```ruby
ai.foo?
ai.bar?
ai.soft?
ai.hard?
end
```

And BTW this is not current behavior of StringInquirer:

```ruby
ActiveSupport::StringInquirer.new(:foo)
# => TypeError: can't convert Symbol into String
```

It only works for Strings and objects which implement `to_str`, which `Symbol` doesn't. And this particular example works on Symbols, so a similar implementation of ArrayInquirer won't work here - presumably, it would raise an exception during ArrayInquirer initialization, before even having a chance to call the method missing.

File: actionpack/lib/action_dispatch/http/mime_negotiation.rb (hunk 9 above)
Comment (kaspth) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: Let's do a deprecation cycle, so people will have an easier time updating.

**Comment by dhh** [comment]
I'm thinking it's a runtime comparison. So it works with both Indifferent and not. I also think that we should have a SymbolInquery. I've wanted that in the past.

**Comment by egilburg** [comment]
@dhh do you prefer two separate inquirer classes, one just for strings and one just for symbols, rather than soften StringInquirer to accept symbols (and just cast them to_s on initialize)?

**Comment by dhh** [comment]
I was thinking about StringInquirer taking a symbol, but then it’s not a symbol any more, it’s a string. I’d rather just be explicit and have a SymbolInquirer that inherits from Symbol.

**Comment by dhh** [comment]
Anyway, let's split those off into their own tickets. Don't want to swamp this ticket with those concerns. I'm very eager to get this in sooner rather than later.

**Comment by georgeclaghorn** [comment]
Update: I've extracted an ArrayInquirer, but need to write tests. My schedule is a little crazy for the time being, but I will get this done.

**Comment by dhh** [comment]
Awesome, George. Not a mega rush. Thanks for helping with this!

File: actionpack/lib/action_dispatch/http/mime_negotiation.rb [hunk 11]
```diff
@@ -139,6 +141,31 @@ def negotiate_mime(order)
...
+        end
+
+        def to_ary
+          @variants
+        end
+
+        private
+          def method_missing(name, *args)
+            if name[-1] == '?'
+              any? name[0..-2].to_sym
```
Comment (egilburg) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: How does this work? `name[0..-2]` will return an array, and arrays don't respond to `to_sym`. Did you mean to `.map(&:to_sym)`?

Comment (egilburg) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: oops nvm

File: actionpack/lib/action_dispatch/http/mime_negotiation.rb (hunk 2 above)
Comment (jvanbaarsen) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: I like the name of this class :+1:

File: actionpack/lib/action_controller/metal/mime_responds.rb [hunk 12]
```diff
@@ -288,7 +288,7 @@ def method_missing(name, *args, &block)
         end
 
         def variant
-          if @variant.nil?
+          if @variant.empty?
             @variants[:none] || @variants[:any]
           elsif (@variants.keys & @variant).any?
             @variant.each do |v|
```
Comment (kaspth) [inline actionpack/lib/action_controller/metal/mime_responds.rb]: Since we're already touching this code, maybe we can improve it. How about this? :smile:

```ruby
def variant
  if @variant.empty?
    @variants[:none] || @variants[:any]
  else
    @variants[ @variant.find { |v| @variants.key?(v) } || :any ]
  end
end
```

**Comment by sdhull** [comment]
Yes please merge this :+1: 

My initial inclination when working with `request.variant` was to inquire like that `request.variant.iphone?`

Failing that, I was surprised/dismayed by the fact `request.variant.include? :iphone` cannot be done unless you explicitly set the variant every time. It seems to me that if you don't set any variants, then `request.variant.should == []`. Actually, I'm not sure this PR actually accomplishes this, however being able to do `request.variant.iphone?` would obviate my need for `request.variant` defaulting to `[]`. (Still strikes me that this would be proper for Principle Of Least Surprise.)

Moreover, since `variant` is not a single value (but rather a list of values), shouldn't it be `request.variants`?

**Comment by georgeclaghorn** [comment]
@sdhull: Agree about pluralizing `variant`, or at least providing a plural alias. Would like to investigate that in a separate PR.

**Comment by sdhull** [comment]
If left unset, will `request.variant == []` be true?

**Comment by georgeclaghorn** [comment]
Updated to use an extracted ArrayInquirer.

@dhh @kaspth

File: actionpack/lib/action_controller/metal/mime_responds.rb (hunk 12 above)
Comment (georgeclaghorn) [inline actionpack/lib/action_controller/metal/mime_responds.rb]: Took it a step further:

```ruby
def variant
  if @variant.empty?
    @variants[:none] || @variants[:any]
  else
    @variants[variant_key]
  end
end

private
  def variant_key
    @variant.find { |variant| @variants.key?(variant) } || :any
  end
```

File: activesupport/lib/active_support/array_inquirer.rb [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+    end
+
+    private
+      def respond_to_missing?(name, include_private = false)
+        name[-1] == '?'
+      end
+
+      def method_missing(name, *args)
+        if name[-1] == '?'
+          any?(name[0..-2]) || any?(name[0..-2].to_sym)
```
Comment (egilburg) [inline activesupport/lib/active_support/array_inquirer.rb]: for consistency between string/symbol, this check should be at the `any?` level:

```ruby
def any?(*candidates, &block)
  candidates.detect do |candidate|
    include?(candidate) || include?(candidate.to_sym)
  end || false
end 

private

def method_missing(name, *args)
  if name[-1] == '?'
    any?(name[0..2])
  end
end
```

File: activesupport/lib/active_support/array_inquirer.rb [hunk 14]
```diff
@@ -0,0 +1,35 @@
+module ActiveSupport
+  # Wrapping an array in an +ArrayInquirer+ gives a friendlier way to check
+  # its string-like contents. For example, <tt>request.variant</tt> returns an
+  # +ArrayInquirer+ object. To check a request's variants, you can call:
+  #
+  #   request.variant.phone?
```
Comment (egilburg) [inline activesupport/lib/active_support/array_inquirer.rb]: We aren't dealing with request here, make explicit usage example:

```ruby
  # variant = ArrayInquirer.new(:phone, :tablet, :desktop)
  #
  # variant.phone? # => true
  # variant.laptop? # => false
  #
  # variant.any?(:phone, :desktop) # => true
  # variant.any?(:phone, :laptop) # => true
  # variant.any?(:laptop, :desktop) # => false
```

File: activesupport/lib/active_support/array_inquirer.rb (hunk 13 above)
Comment (kaspth) [inline activesupport/lib/active_support/array_inquirer.rb]: Your suggested code changes are a little off. We don't need to check for `none?` or explicitly return false. But I agree with moving the check :+1:

File: actionpack/lib/action_controller/metal/mime_responds.rb (hunk 12 above)
Comment (kaspth) [inline actionpack/lib/action_controller/metal/mime_responds.rb]: Great, :+1:

File: activesupport/lib/active_support/array_inquirer.rb (hunk 13 above)
Comment (georgeclaghorn) [inline activesupport/lib/active_support/array_inquirer.rb]: Fixed.

File: activesupport/lib/active_support/array_inquirer.rb (hunk 14 above)
Comment (georgeclaghorn) [inline activesupport/lib/active_support/array_inquirer.rb]: Fixed.

File: activesupport/CHANGELOG.md [hunk 15]
```diff
@@ -1,3 +1,23 @@
...
+
+        variants = ActiveSupport::ArrayInquirer.new([:phone, :tablet])
+
+        variants.phone?    # => true
+        variants.tablet?   # => true
+        variants.desktop?  # => false
+
+        variants.any?(:phone, :tablet)   # => true
+        variants.any?(:phone, :desktop)  # => true
+        variants.any?(:desktop, :watch)  # => false
```
Comment (kaspth) [inline activesupport/CHANGELOG.md]: Clever thinking with the watch here :smile:

**Comment by kaspth** [comment]
Looks great to me, :+1:

File: activesupport/test/array_inquirer_test.rb [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+    assert @array_inquirer.any?(:watch, :tablet)
+    assert_not @array_inquirer.any?(:desktop, :watch)
+  end
+
+  def test_any_with_block
+    assert @array_inquirer.any? { |v| v == :mobile }
+    assert_not @array_inquirer.any? { |v| v == :desktop }
+  end
+
+  def test_respond_to
```
Comment (egilburg) [inline activesupport/test/array_inquirer_test.rb]: assert the shortcut method since you added it:

```ruby
def test_inquiry
  assert_equal @array_inquirer, [:mobile, :tablet].inquiry
end
```

**Comment by rafaelfranca** [comment]
Thank you so much. The code is great, but I'd not include the `Array#inquiry` we are promoting too much a feature that I don't think it will be used in most part of the applications. Lets keep just the `ArrayInquirer`  constructor.

**Comment by dhh** [comment]
Great work, George! :metal:

**Comment by dhh** [comment]
I think the ArrayInquirer makes as much sense as the StringInquirer. I'm certainly going to use it. I think it's fair to highlight that it's there.

**Comment by rafaelfranca** [comment]
I mean, it will still be highlighted in the CHANGELOG and documentation but I don't think it needs a `Array#inquiry` method. Even that we have a lot of usage of this feature it will be usually well encapsulated, so we can call `ActiveSupport::ArrayInquirer.new` without losing the legibility.

**Comment by dhh** [comment]
I like the #inquiry form, as it makes it possible to use this inline on a passed parameter without extra setup. And it then maps directly to String#inquiry as well.

**Comment by rafaelfranca** [comment]
:+1:, so I'll add back

**Comment by rafaelfranca** [comment]
Add back at b5c35023bd422c62973d003378b7185dfe221919
